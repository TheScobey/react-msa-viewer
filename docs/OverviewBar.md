# `OverviewBar` (component)

Creates a small overview box of the sequences for a general overview.

## Props

### `fillColor`

defaultValue: `"#999999"`


### `height`

defaultValue: `50`


### `method`

defaultValue: `"conservation"`

