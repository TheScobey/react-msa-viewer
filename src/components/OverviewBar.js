/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React, { Component } from 'react';
import msaConnect from '../store/connect'

import PropTypes from 'prop-types';

import { throttle } from 'lodash-es';

import Canvas from '../drawing/canvas';

import createRef from 'create-react-ref/lib/createRef';

const MSAStats = require('stat.seqs');

class OverviewBarComponent extends Component {

  constructor(props) {
    super(props);
    this.canvas = createRef();
    this.draw = throttle(this.draw, this.props.viewpoint.msecsPerFps);
    this.calculateStats();
  }

  componentDidMount() {
    this.ctx = new Canvas(this.canvas.current);
    this.draw();
  }

  componentDidUpdate() {
    this.calculateStats();
    this.draw();
  }

  draw() {
    this.ctx.startDrawingFrame();
    this.ctx.font(this.props.viewpoint.positionFont);

    const tileWidth = this.props.viewpoint.tileSizes[0];
    const yPos = 0;
    const startTile = Math.floor(this.props.position.xPos / tileWidth);
    const tiles = Math.ceil(this.props.viewpoint.width / tileWidth) + 1;
    let xPos = -this.props.position.xPos % tileWidth;
    for (let i = startTile; i < (startTile + tiles); i++) {
			let height = this.props.height * this.columnHeights[i];
			const remainingHeight = this.props.height - height;
      this.ctx.fillStyle(this.props.fillColor);
      this.ctx.fillRect(xPos, yPos + remainingHeight, tileWidth, height);
      xPos += this.props.viewpoint.tileSizes[0];
    }
    this.ctx.endDrawingFrame();
  }

  // TODO: do smarter caching here
  calculateStats() {
    const stats = MSAStats(this.props.sequences.map(e => e.sequence));
    this.columnHeights = [];
    switch (this.props.method) {
      case "conservation":
        this.columnHeights = stats.scale(stats.conservation());
        break;
      case "information-content":
        this.columnHeights = stats.scale(stats.ic());
        break;
      default:
        console.error(this.props.method + "is an invalid aggregation method for <OverviewBar />");
    }
  }

  render() {
    return (
      <canvas
        ref={this.canvas}
        width={this.props.viewpoint.width}
        height={this.props.height}
      />
    );
  }
}

OverviewBarComponent.defaultProps = {
  height: 50,
  fillColor: "#999999",
  method: "conservation",
}

OverviewBarComponent.PropTypes = {
  /**
   * Method to use for the OverviewBar:
   *  - `information-content`: Information entropy after Shannon of a column (scaled)
   *  - `conservation`: Conservation of a column (scaled)
   */
  method: PropTypes.oneOf('information-content', 'conservation'),

  /**
   * Height of the OverviewBar (in pixels), e.g. `100`
   */
  height: PropTypes.number,

  /**
   * Fill color of the OverviewBar, e.g. `#999999`
   */
  fillColor: PropTypes.string,
};

const mapStateToProps = state => {
  return {
    sequences: state.sequences.raw,
    position: state.position,
    viewpoint: state.viewpoint,
  }
}

export default msaConnect(
  mapStateToProps,
  //mapDispatchToProps
)(OverviewBarComponent);
