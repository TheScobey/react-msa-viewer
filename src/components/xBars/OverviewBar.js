/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/
import React, { Component, PureComponent } from 'react';
import PropTypes from 'prop-types';
import shallowCompare from 'react-addons-shallow-compare';

import {
  pick
} from 'lodash-es';

import XBar from './xBar';
import msaConnect from '../../store/connect'
import MSAStats from '../../utils/statSeqs';

function createBar({columnHeights, tileWidth, height, fillColor,
  barStyle, barAttributes}) {
  class Bar extends PureComponent {
    render() {
      const { index, ...otherProps} = this.props;
      otherProps.style = {
        height: Math.round(columnHeights[index] * height),
        width: tileWidth,
        display: "inline-block",
        textAlign: "center",
        backgroundColor: fillColor,
      }
      return (
        <div {...otherProps}>
        </div>
      );
    }
  }
  return Bar;
}

/**
 * Creates a small overview box of the sequences for a general overview.
 */
class HTMLOverviewBarComponent extends Component {

  componentWillMount() {
    this.calculateStats();
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (["height", "sequences", "tileWidth"].some(key=> {
      return nextProps[key] !== this.props[key];
    }, true)){
      if (nextProps["sequences"] !== this.props.sequences) {
        this.calculateStats();
      }
      return true;
    }
    return shallowCompare(this, nextProps, nextState);
  }

  // TODO: do smarter caching here (reselect)
  calculateStats() {
    const stats = MSAStats(this.props.sequences.map(e => e.sequence));
    let columnHeights;
    switch (this.props.method) {
      case "conservation":
        columnHeights = stats.scale(stats.conservation());
        break;
      case "information-content":
        columnHeights = stats.scale(stats.ic());
        break;
      default:
        console.error(this.props.method + "is an invalid aggregation method for <OverviewBar />");
    }

    this.statsBar = createBar({
      columnHeights: columnHeights,
      ...pick(this.props, [
        "tileWidth", "height", "fillColor",
        "barStyle", "barAttributes",
      ])
    });
  }

  render() {
    const {cacheElements,
      height,
      method,
      fillColor,
      dispatch,
      barStyle,
      barAttributes,
      ...otherProps} = this.props;
    return (
      <XBar
        tileComponent={this.statsBar}
        cacheElements={cacheElements}
        {...otherProps}
      />
    );
  }
}

HTMLOverviewBarComponent.defaultProps = {
  height: 50,
  fillColor: "#999999",
  method: "conservation",
  cacheElements: 10,
}

HTMLOverviewBarComponent.propTypes = {
  /**
   * Method to use for the OverviewBar:
   *  - `information-content`: Information entropy after Shannon of a column (scaled)
   *  - `conservation`: Conservation of a column (scaled)
   */
  method: PropTypes.oneOf(['information-content', 'conservation']),

  /**
   * Height of the OverviewBar (in pixels), e.g. `100`
   */
  height: PropTypes.number,

  /**
   * Fill color of the OverviewBar, e.g. `#999999`
   */
  fillColor: PropTypes.string,

  /**
   * Inline styles to apply to the OverviewBar component
   */
  style: PropTypes.object,

  /**
   * Inline styles to apply to each bar.
   */
  barStyle: PropTypes.object,

  /**
   * Attributes to apply to each bar.
   */
  barAttributes: PropTypes.object,
};

const mapStateToProps = state => {
  return {
    sequences: state.sequences.raw,
    maxLength: state.sequences.maxLength,
    width: state.props.width,
    tileWidth: state.props.tileWidth,
    nrXTiles: state.sequenceStats.nrXTiles,
  }
}

export default msaConnect(
  mapStateToProps,
)(HTMLOverviewBarComponent);
// for testing
export {
  HTMLOverviewBarComponent as OverviewBar,
}
