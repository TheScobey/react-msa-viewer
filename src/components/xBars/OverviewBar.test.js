/**
* Copyright 2018, Plotly, Inc.
* All rights reserved.
*
* This source code is licensed under the MIT license found in the
* LICENSE file in the root directory of this source tree.
*/

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import {
  dummySequences,
  FakePositionStore,
} from '../../test';

import { OverviewBar } from './OverviewBar';

it('renders properly', () => {
  const component = renderer.create(
    <FakePositionStore currentViewSequencePosition={0}>
      <OverviewBar nrXTiles={5} tileWidth={20} width={200} maxLength={100} sequences={[...dummySequences]} />
    </FakePositionStore>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders properly with a moved viewport', () => {
  const div = document.createElement('div');
  ReactDOM.render(<FakePositionStore currentViewSequencePosition={40}>
      <OverviewBar nrXTiles={5} tileWidth={20} width={200} maxLength={100} sequences={[...dummySequences]} />
    </FakePositionStore>, div);
  ReactDOM.unmountComponentAtNode(div);
});
