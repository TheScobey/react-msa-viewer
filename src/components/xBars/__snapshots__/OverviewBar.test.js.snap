// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<div
  style={
    Object {
      "height": undefined,
    }
  }
>
  <div
    style={
      Object {
        "overflow": "hidden",
        "position": "relative",
        "whiteSpace": "nowrap",
        "width": 200,
      }
    }
  >
    <div>
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 34,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 32,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 35,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 28,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 21,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 35,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 50,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "#999999",
            "display": "inline-block",
            "height": 35,
            "textAlign": "center",
            "width": 20,
          }
        }
      />
    </div>
  </div>
</div>
`;
